### PR TITLE
PIM-4016: Be able to use Simple Reference Data as Variant Group Axis

### PR DESCRIPTION
**Pull request by skeleton**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | Y
| BC breaks?           | Y
| CI currently passes? | ?
| Tests pass?          | Y
| Scenarios pass?      | Y
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   |  N
| Fixed tickets        | PIM-4016
| DB schema updated?   | N
| Migration script?    | N
| Doc PR               | N

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php [hunk 1]
```diff
@@ -60,6 +64,16 @@ public function setAttributeRepository(AttributeRepositoryInterface $attributeRe
     /**
      * {@inheritdoc}
      */
+    public function setReferenceDataRegistry(ConfigurationRegistryInterface $registry)
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php]: why an injection by setter ?

File: src/Pim/Bundle/CatalogBundle/Resources/config/storage_driver/doctrine/orm.yml [hunk 2]
```diff
@@ -63,6 +63,7 @@ services:
         calls:
             - [setProductQueryBuilderFactory, ['@pim_catalog.query.product_query_builder_factory']]
             - [setAttributeRepository, ['@pim_catalog.repository.attribute']]
+            - [setReferenceDataRegistry, ['@pim_reference_data.registry']]
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Resources/config/storage_driver/doctrine/orm.yml]: why an injection by setter ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php [hunk 3]
```diff
@@ -379,39 +393,36 @@ public function valueExists(ProductValueInterface $value)
...
+            'INNER JOIN pim_catalog_group_type gt on gt.id = g.type_id ' .
+            'WHERE ga.group_id = :groupId AND gt.code = "VARIANT" ' .
+            'AND (v.option_id IS NOT NULL';
+
+        $references = $this->referenceDataRegistry->all();
+        if (!empty($references)) {
+            $valueMetadata = QueryBuilderUtility::getProductValueMetadata($this->_em, $this->_entityName);
+
+            foreach ($references as $code => $referenceData) {
+                if ('simple' === $referenceData->getType()) {
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php]: use the constant

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php [hunk 4]
```diff
@@ -379,39 +393,36 @@ public function valueExists(ProductValueInterface $value)
...
+                    if ($valueMetadata->isAssociationWithSingleJoinColumn($code)) {
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php]: `null !== ` or `false !== ` ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php [hunk 5]
```diff
@@ -379,39 +393,36 @@ public function valueExists(ProductValueInterface $value)
...
-        $sql = sprintf(
-            'SELECT * FROM (%s) as p WHERE p.product_id NOT IN (%s);',
-            $elligibleProductsSQL,
-            $alreadyInGroupSQL
-        );
-        $sql = QueryBuilderUtility::prepareDBALQuery($this->_em, $this->_entityName, $sql);
+        $sql = 'SELECT v.entity_id AS product_id, gp.group_id, ga.group_id ' .
+            'FROM pim_catalog_group g ' .
+            'INNER JOIN pim_catalog_group_attribute ga ON ga.group_id = g.id ' .
+            'INNER JOIN %product_value_table% v ON v.attribute_id = ga.attribute_id ' .
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php]: can you do the same with other tables ?

yeah, I know it was like that before :)

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php (hunk 4 above)
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php]: nop, it's a boolean, it's not required

**Pull request merged**